### PR TITLE
VMKernelCrashError:Add a new exception type

### PR DESCRIPTION
**Pull request by zhenyzha**
Add a new exception type, which is used to describe the
call trace of warning type, and this type of call trace will
not cause VM dead. But it will cause some codes to crash inside
the kernel, so a new function is needed to describe it.

ID: 2178511
Signed-off-by: zhenyzha [zhenyzha@redhat.com](mailto:zhenyzha@redhat.com)

### CODE REVIEW

**Comment by zhenyzha** [comment]
@PaulYuuu @YongxueHong Could you pls help review when you are free? thanks.

```
 (1/5) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.numa_opts.nodes.0.arm64-pci: STARTED
 (1/5) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.numa_opts.nodes.0.arm64-pci: PASS (47.31 s)
 (2/5) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.numa_opts.nodes.1.arm64-pci: STARTED
 (2/5) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.numa_opts.nodes.1.arm64-pci: PASS (41.31 s)
 (3/5) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.numa_opts.nodes.2.arm64-pci: STARTED
 (3/5) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.numa_opts.nodes.2.arm64-pci: ERROR: VM hit a kernel crash, see debug/serial log for details (44.85 s)
 (4/5) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.numa_opts.nodes.3.arm64-pci: STARTED
 (4/5) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.numa_opts.nodes.3.arm64-pci: PASS (44.64 s)
 (5/5) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.numa_opts.nodes.64.arm64-pci: STARTED
 (5/5) Host_RHEL.m9.u3.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.2.0.aarch64.io-github-autotest-qemu.numa_opts.nodes.64.arm64-pci: ERROR: VM hit a kernel crash, see debug/serial log for details (172.27 s)
```

File: virttest/virt_vm.py [hunk 1]
```diff
@@ -1002,7 +1012,7 @@ def verify_kernel_crash(self):
                 return
             match = re.search(panic_re, data, re.DOTALL | re.MULTILINE | re.I)
             if match:
-                raise VMDeadKernelCrashError(match.group(0))
+                raise VMKernelCrashError(match.group(0))
```
Comment (PaulYuuu) [inline virttest/virt_vm.py:1015]: This change will cause some test case checkpoints to be failed, so with this change, we need to update test case part, not only tp-qemu, but also tp-libvirt.

Comment (zhenyzha) [reply virttest/virt_vm.py:1015]: About tp-qemu and tp-libvirt cases failed.
I'm not sure if this step should be modified at our discretion.
This is more like a product issue causing automation cases to fail, 
and ultimately whether to keep this checkpoint may be up to the feature owner to decide.

Comment (zhenyzha) [reply virttest/virt_vm.py:1015]: According to the offline discussion, I will send a depend patch 
to update the lower layer call about the exception.
thanks for reminding :)

Comment (harihare) [reply virttest/virt_vm.py:1015]: VMDeadKernelCrashError handles both Kernel trace and crash. Then why the change required?

Comment (zhenyzha) [reply virttest/virt_vm.py:1015]: > VMDeadKernelCrashError handles both Kernel trace and crash. Then why the change required?

Please refer to the patch for the background: https://github.com/avocado-framework/avocado-vt/pull/3635

Comment (zhenyzha) [reply virttest/virt_vm.py:1015]: Thanks a lot, I think your suggestion is very beneficial. 
After matching we can refine exception types. 
I'll commit the changes later.

**Comment by zhenyzha** [comment]
@PaulYuuu @YongxueHong Could you pls help review again when you are free? thanks.

File: virttest/virt_vm.py [hunk 2]
```diff
@@ -1002,7 +1012,10 @@ def verify_kernel_crash(self):
                 return
             match = re.search(panic_re, data, re.DOTALL | re.MULTILINE | re.I)
             if match:
-                raise VMDeadKernelCrashError(match.group(0))
+                if "BUG" in match.group(0):
```
Comment (YongxueHong) [inline virttest/virt_vm.py:1015]: I don't think that we can judge the VM as dead according to ` if "BUG" in match.group(0)`.
I think we check the status of the VM and then raise the corresponding error.

Comment (zhenyzha) [reply virttest/virt_vm.py]: Thanks for the reminder, it has been corrected.

File: virttest/virt_vm.py [hunk 3]
```diff
@@ -1002,7 +1012,10 @@ def verify_kernel_crash(self):
                 return
             match = re.search(panic_re, data, re.DOTALL | re.MULTILINE | re.I)
             if match:
-                raise VMDeadKernelCrashError(match.group(0))
+                if self.is_dead():
+                    raise VMDeadKernelCrashError(match.group(0))
+                else:
+                    raise VMKernelCrashError(match.group(0))
```
Comment (YongxueHong) [inline virttest/virt_vm.py:1018]:
```suggestion
                raise VMKernelCrashError(match.group(0))
```

File: virttest/virt_vm.py [hunk 4]
```diff
@@ -1002,7 +1012,10 @@ def verify_kernel_crash(self):
                 return
             match = re.search(panic_re, data, re.DOTALL | re.MULTILINE | re.I)
             if match:
-                raise VMDeadKernelCrashError(match.group(0))
+                if self.is_dead():
```
Comment (PaulYuuu) [inline virttest/virt_vm.py:1015]: `is_dead` can only check the qemu process, not the guest status. when the guest crashes, if qemu can report `guest-panicked` status, then we can use it, but if not, there is not a good idea the check the guest status.

Comment (YongxueHong) [reply virttest/virt_vm.py:1015]: > `is_dead` can only check the qemu process, not the guest status. when the guest crashes, if qemu can report `guest-panicked` status, then we can use it, but if not, there is not a good idea the check the guest status.

Hi @PaulYuuu 
You's concern makes sense. 
I think we should give the definition of the `VMDead` of `VMDeadKernelCrashError`.
What's the meaning of `VMDead`? Does it relate to the VM class?
In another way, how do we define the VM's dead status here?
CC @zhenyzha

File: virttest/virt_vm.py (hunk 3 above)
Comment (zhenyzha) [reply virttest/virt_vm.py]: Corrected

File: virttest/virt_vm.py [hunk 5]
```diff
@@ -988,8 +998,8 @@ def verify_kernel_crash(self):
         """
         Find kernel crash message on the VM serial console.
 
-        :raise: VMDeadKernelCrashError, in case a kernel crash message was
-                found.
+        :raise: VMKernelCrashError or VMDeadKernelCrashError, in case
+                a kernel crash message was found.
```
[the remark this answers is not in the document]
Comment (zhenyzha) [reply virttest/virt_vm.py]: Corrected

File: virttest/virt_vm.py (hunk 4 above)
Comment (zhenyzha) [reply virttest/virt_vm.py:1016]: Hello guys,

For me, it's relatively accurate to use it for checking in `virt_vm`.
We can refer to `verify_alive(self)`, its judgment is also through `self.is_dead()`
https://github.com/avocado-framework/avocado-vt/blob/master/virttest/virt_vm.py#L699

If we really want to explicitly define `VMDead` from the abstraction layer, 
I think overall refactoring may be needed here.
Of course, if there is a better way of judging, I am also happy to correct it.

File: virttest/virt_vm.py [hunk 6]
```diff
@@ -185,6 +185,16 @@ def __str__(self):
...
+
+    def __init__(self, kernel_crash):
+        VMError.__init__(self, kernel_crash)
+        LOG.debug(kernel_crash)
+
+    def __str__(self):
+        return ("VM hit a kernel crash, see debug/serial log for details")
+
+
 class VMDeadKernelCrashError(VMError):
```
Comment (PaulYuuu) [inline virttest/virt_vm.py:198]:
```suggestion
class VMDeadKernelCrashError(VMKernelCrashError):
```

File: virttest/virt_vm.py (hunk 4 above)
Comment (PaulYuuu) [reply virttest/virt_vm.py:1016]: I think you should check `verify_alive` in qemu_vm.py and libvirt_vm.py, this method in virt_vm.py never be used for tp-qemu and tp-libvirt.

**Comment by zhenyzha** [comment]
After discussion, we think that the verification method of the verify_kernel_crash function itself is inaccurate. On the other hand, the definition of VMDead is different in tp-libvirt and tp-qemu. So finally decided to close this patch. 
If we want to make changes in the future, we may need to refactor this function as a whole from the definition.

**Pull request closed without merging**